### ISSUE TITLE
No compiler complaints with fat arrow followed by squiggly brackets 

### ISSUE THREAD

**Issue by CP-casp** [dart-lang/sdk#47280]
This tracker is for issues related to: (I'm not sure)

* Analyzer
* Dart core libraries ("dart:async", "dart:io", etc.)
* Dart VM
* dart2js
* dev_compiler

Dart SDK version: 2.14.2 (stable) (Wed Sep 15 12:32:06 2021 +0200) on "windows_x64"

I have recently upgraded flutter and dart to the latest, but tested this with downgraded versions as well (via 'flutter downgrade')

There are instances in my code where I can use functions that have fat arrow and squiggly brackets. I can breakpoint in them and hit those breakpoints as if nothing is wrong. In other instances the code wont even compile.

Example code snippet:

```dart
void registerAll(List behaviors) => {
      for (var behavior in behaviors)
        {
//           behavior.initialize(),
          print("List item content: $behavior"),
        }
    };

void main() {
  registerAll(["item1", "item2", "item3"]);
}
```

From reading online and looking at a number of tutorials I shouldn't be able to do this. It has also resulted in some very strange behaviour in my app, no crashing, just very hard to follow debugging experience.

**Comment by TimWhiting** [comment]
```dart
void registerAll(List behaviors) => <Set<void>>{
      for (var behavior in behaviors)
        <void>{
//           behavior.initialize(),
          print("List item content: $behavior"),
        }
    };
```
I believe this is inferring a set literal containing a set.

**Comment by lrhn** [comment]
Correct. The *comma* after the `print` call is what's causing it to be a valid set-of-set-of-void literal. 
If it was intended as statement blocks, that should have been a `;`.
